tests/test_readonly_views.py

```diff
@@ -0,0 +1,223 @@
+    passed in _assert_refcount has 3 extra refcounts:
+        - From the obj variable
+        - From passing in _assert_refcount
+        - From passing in sys.getrefcount
+    """
+    actual_refcount = sys.getrefcount(obj) - 3
+    assert actual_refcount == expected_refcount, (
+        f"\nActual:   {actual_refcount}\n"
+        f"Expected: {expected_refcount}"
+    )
+
+
+@pytest.mark.parametrize("MorphologyClass, path", READONLY_CLASSES_PATHS)
+def test_morphology_level_immutability(MorphologyClass, path):
+
+    morph = MorphologyClass(path)
+
+    # ensure we cannot overwrite the attributes
+    with pytest.raises(AttributeError):
+        morph.points = np.ones(morph.points.shape)
+
+    with pytest.raises(AttributeError):
+        morph.diameters = np.ones(morph.diameters.shape)
+
+    with pytest.raises(AttributeError):
+        morph.section_types = np.ones(morph.section_types.shape)
+
+    # ensure we cannot change the underlying data
+    with pytest.raises(ValueError):
+        morph.points[:] = np.ones(morph.points.shape)
+
+    with pytest.raises(ValueError):
+        morph.diameters[:] = np.ones(morph.diameters.shape)
+
+    with pytest.raises(ValueError):
+        morph.section_types[:] = np.ones(morph.section_types.shape)
+
+    # check if flag set correctly
+    assert not morph.points.flags["WRITEABLE"]
+    assert not morph.diameters.flags["WRITEABLE"]
+    assert not morph.section_types.flags["WRITEABLE"]
+
+    # check that the parent of the array is the morphology
+    assert morph.points.base is morph
+    assert morph.diameters.base is morph
+    assert morph.section_types.base is morph
+
+    # make the same tests when the datasets are assigned to a variable
+    points = morph.points
+    _assert_refcount(points, 1)
+    assert points.base is morph
+    assert points.flags["WRITEABLE"] == False
+
+    diameters = morph.diameters
+    _assert_refcount(diameters, 1)
+    assert diameters.base is morph
+    assert diameters.flags["WRITEABLE"] == False
+
+    section_types = morph.section_types
+    _assert_refcount(section_types, 1)
+    assert section_types.base is morph
+    assert section_types.flags["WRITEABLE"] == False
+
+
+@pytest.mark.parametrize("MorphologyClass, path", READONLY_CLASSES_PATHS)
+def test_section_level_immutability__iter(MorphologyClass, path):
+
+    morph = MorphologyClass(path)
+    _assert_refcount(morph, 1)
+
+    for section in morph.iter():
+
+        _assert_refcount(section, 1)
+        # 1 (morph) + 1 (iter())
+        _assert_refcount(morph, 2)
+
+        points = section.points
+        assert points.base is section
+        _assert_refcount(section, 2)
+        _assert_refcount(morph, 2)
+
+        diameters = section.diameters
+        assert diameters.base is section
+        _assert_refcount(section, 3)
+        _assert_refcount(morph, 2)
+
+        #del points, diameters
+        #_assert_refcount(section, 1)
```

Comment:
The Vasculature's last section refcount is 3 instead of 1 if the points & diameters are not cleared. Very peculiar...